serialization.rst

```diff
@@ -0,0 +1,337 @@
+
+    The rationale for the latter is because no explicit choice about whether
+    to use ODM features has been made, it should not bail out because of an
+    extra ``__pclass`` property. The behaviour is different if the data type
+    was mapped to ``"odm"`` (see below).
+
+- ``"array"`` — turns a BSON array or BSON document into a PHP array.
+  ``__pclass`` properties [1]_ are igored.
+
+- ``"object"`` or ``"stdClass"`` — turns a BSON array or BSON document into a
+  ``stdClass`` object. There will be no special treatment of a ``__pclass``
+  property [1]_, buy it should **not** be set as property in the returned
+  object.
+
+- ``"odm"`` — uses the class name set in the ``__pclass`` property of a BSON
+  array or BSON document to determine the class name of the created object.
+
+  The properties of the BSON document, minus the ``__pclass`` property are
+  sent as an associative array to the ``bsonUnserialize`` function to initialise
+  the object's properties.
+
+  Remarks:
+
+  - If the ``__pclass`` property is not present [1]_, this throws the
+    ``MongoDB\Driver\Exception\UnexpectedValueException`` exception with a
+    message indicating the special class marker could not be found.
+
+  - If the class as determined by the ``__pclass`` property does not implement
+    the ``MongoDB\BSON\Persistable`` interface, throw an
+    ``MongoDB\Driver\Exception\UnexpectedValueException`` exception with a
+    message indicating that the named class does not implement the required
+    interface.
+
+- ``any other string`` — defines the class name that the BSON array or BSON
+  object should be deserialized at.
+
+  If the class implements the ``MongoDB\BSON\Unserializable`` interface, then
+  the properties of the BSON document, **minus** the ``__pclass`` property [1]_
+  if it exists, are sent as an associative array to the ``bsonUnserialize``
+  function to initialise the object's properties.
+
+  To discuss:
+
+  - What should we do if the named class implements
+    ``MongoDB\BSON\Persistable``?
+    
+    We can throw an ``MongoDB\Driver\Exception\UnexpectedValueException``
+    exception indicating that the BSON document really knows what PHP type it
+    should be. This could be important if the object needs to be deserialized
+    in a special way.
+
+    We can ignore it and act as the ``__pclass`` property wasn't set at all.
+    And not send the ``__pclass`` property as an array element to
+    ``bsonUnserialize``.
+
+TypeMaps
+--------
+
+TypeMaps can be set through the ``setTypeMap()`` on a
+``MongoDB\Driver\Cursor`` object, or the ``$typeMap`` argument of
+``MongoDB\BSON\toPHP()`` (previously, ``MongoDB\BSON\toArray()``). Each of the
+three classes (``root``, ``document`` and ``array``) can be individually set.
+If ``document`` is set to something else than the default and ``root`` is set
+to the default (or not set at all), then the value of ``document`` also counts
+for ``root``. Which makes the following example equivalent::
```

Comment:
I agree. I had understood from @jmikola that this is something you suggested. I'll remove it.